core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -524,6 +529,15 @@ class SparkContext(config: SparkConf) extends Logging {
     executorEnvs ++= _conf.getExecutorEnv
     executorEnvs("SPARK_USER") = sparkUser
 
+    _shuffleDriverComponents = ShuffleDataIOUtils.loadShuffleDataIO(config).driver()
+    _shuffleDriverComponents.initializeApplication().asScala.foreach { case (k, v) =>
+      val key = ShuffleDataIOUtils.SHUFFLE_SPARK_CONF_PREFIX + k
+      if (_conf.contains(key)) {
```

Comment:
Hmmm, isn't this preventing the config from being added? Maybe what you're looking after is `setIfMissing`?

I saw some comments about users being able to set these. I don't think that's the goal here. These configs shouldn't be user-settable; if there are user-settable configs, they should not come from this set of configs, but from normal configs you get directly from `SparkConf`.

For that reason, I'd even change the prefix here so make it clearer this is an internal thing (e.g. call it `spark.shuffle.plugin.__conf__`).

Also, all that kinda tells me you're missing a unit test.

Reply:
I wasn't really saying that users *should* be able to set it, I was just wondering what the behavior would be if users happened to set it.  Are the things they set silently ignored, do they take precedence, is there a warning or a hard failure, or is the behavior undefined?  It sounds like you're saying we should discourage them from doing it (by naming), with the plugin overwriting anything the user speficies.  That is fine by me, just wanted to discuss it a bit.

Reply:
So the consensus is that the plugin overrides anything that the user sets, and we log a warning whenever that happens?

Reply:
My suggestion is to not even bother with trying to detect anything. Write the configs from the driver plugin, read everything on the executor side, don't warn about anything.

This should be considered an internal namespace (which is why I suggested that prefix) and users shouldn't mess with it. If they do, and they run into problems, it's their problem.